### PR TITLE
[QHC-524] Implement measure for Qblox and QM

### PR DESCRIPTION
**Pull request by fedonman**

### CODE REVIEW

File: docs/releases/changelog-dev.md [hunk 1]
```diff
@@ -51,7 +51,18 @@
         Q: !Square {amplitude: 1.0, duration: 200}
   ```
 
+  Calibrated waveforms and weights can be used in QProgram by providing their name.
+
+  ```Python
+  qp = QProgram()
+  qp.play(bus='drive_q0_bus', waveform='Xpi')
+  qp.measure(bus='readout_q0_bus', waveform='Measure', weights='optimal_weights')
```
[the remark this answers is not in the document]
Comment (fedonman) [reply docs/releases/changelog-dev.md:59]: This is the example for the calibration though. I added a changelog entry in the last PR that you approved for the `measure` method specifically, that includes the `save_adc`.

File: src/qililab/qprogram/operations/measure.py [hunk 2]
```diff
@@ -22,25 +22,47 @@
 class Measure(Operation):  # pylint: disable=missing-class-docstring
     bus: str
     waveform: IQPair
-    weights: IQPair | tuple[IQPair, IQPair] | tuple[IQPair, IQPair, IQPair, IQPair] | None
-    demodulation: bool
-    save_raw_adc: bool
+    weights: IQPair
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/qprogram/operations/measure.py:25]: That's the point of the whole PR, to treat weights as a single IQPair (one weight for I and one weight for Q). The 4 matrix elements is just how QM handles the maths internally. There is no valid combination you cannot create by providing `weights` and `rotation`.

File: src/qililab/qprogram/qprogram.py [hunk 3]
```diff
@@ -653,3 +682,223 @@ class _AverageContext(_BlockContext):  # pylint: disable=too-few-public-methods
...
+            demodulation: bool = True,
+        ):
+            """Play a pulse and acquire results.
+
+            Args:
+                bus (str): Unique identifier of the bus.
+                waveform (IQPair): Waveform played during measurement.
+                weights (IQPair): Weights used during demodulation/integration.
+                save_adc (bool, optional): If ADC data should be saved. Defaults to False.
+                demodulation (bool, optional): If demodulation is enabled. Defaults to True.
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/qprogram/qprogram.py:785]: fixed

File: src/qililab/qprogram/qprogram.py [hunk 4]
```diff
@@ -653,3 +682,223 @@ class _AverageContext(_BlockContext):  # pylint: disable=too-few-public-methods
...
+            demodulation: bool = True,
+        ):
+            """Play a named pulse and acquire results.
+
+            Args:
+                bus (str): Unique identifier of the bus.
+                waveform (str): Waveform played during measurement.
+                weights (IQPair): Weights used during demodulation/integration.
+                save_adc (bool, optional): If ADC data should be saved. Defaults to False.
+                demodulation (bool, optional): If demodulation is enabled. Defaults to True.
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/qprogram/qprogram.py:806]: fixed

File: src/qililab/qprogram/qprogram.py [hunk 5]
```diff
@@ -653,3 +682,223 @@ class _AverageContext(_BlockContext):  # pylint: disable=too-few-public-methods
...
+            demodulation: bool = True,
+        ):
+            """Play a named pulse and acquire results.
+
+            Args:
+                bus (str): Unique identifier of the bus.
+                waveform (str): Waveform played during measurement.
+                weights (str): Weights used during demodulation/integration.
+                save_adc (bool, optional): If ADC data should be saved. Defaults to False.
+                demodulation (bool, optional): If demodulation is enabled. Defaults to True.
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/qprogram/qprogram.py:848]: fixed

File: src/qililab/qprogram/qprogram.py [hunk 6]
```diff
@@ -653,3 +682,223 @@ class _AverageContext(_BlockContext):  # pylint: disable=too-few-public-methods
...
+            demodulation: bool = True,
+        ):
+            """Play a named pulse and acquire results.
+
+            Args:
+                bus (str): Unique identifier of the bus.
+                waveform (IQPair): Waveform played during measurement.
+                weights (str): Weights used during demodulation/integration.
+                save_adc (bool, optional): If ADC data should be saved. Defaults to False.
+                demodulation (bool, optional): If demodulation is enabled. Defaults to True.
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/qprogram/qprogram.py:827]: fixed

File: src/qililab/qprogram/qprogram.py [hunk 7]
```diff
@@ -653,3 +682,223 @@ class _AverageContext(_BlockContext):  # pylint: disable=too-few-public-methods
...
+            rotation: float = 0.0,
+            demodulation: bool = True,
+        ):
+            """Play a pulse and acquire results.
+
+            Args:
+                bus (str): Unique identifier of the bus.
+                waveform (IQPair): Waveform played during measurement.
+                weights (IQPair): Weights used during demodulation/integration.
+                rotation (float): Angle in radians to rotate the IQ plane during demodulation/integration.
```
[the remark this answers is not in the document]
Comment (fedonman) [reply src/qililab/qprogram/qprogram.py]: fixed

File: docs/releases/changelog-dev.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (jordivallsq) [reply docs/releases/changelog-dev.md:59]: You are right!

File: src/qililab/qprogram/operations/measure.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jordivallsq) [reply src/qililab/qprogram/operations/measure.py:25]: Perfect, if `weights `+ `rotation `allows us to implement the same functionality as a matrix we can work with it without losing the QM advantages.

**Comment by fedonman** [comment]
@jordivallsq If everything is ok, approve this as well, because it's the first in the line to merge.

**Review by jordivallsq — APPROVED** [review]
Gret job!

**Pull request merged**